python/triton/compiler/compiler.py

```diff
@@ -156,7 +156,12 @@ def triton_key():
     for lib in pkgutil.iter_modules([compiler_path, backends_path]):
         with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
             contents += [hashlib.sha256(f.read()).hexdigest()]
-    # backend
+    # Python backend
+    tp_backends_path = os.path.join(TRITON_PATH, 'backends')
```

Comment:
Why isn't this part of the existing iter_modules call above?

Reply:
I tried but that lost some files. I suspect it's a bug of `pkgutil.iter_modules`. 

Fused into exising iter_modules:
<_io.BufferedReader name='python/triton/compiler/code_generator.py'>
<_io.BufferedReader name='python/triton/compiler/compiler.py'>
<_io.BufferedReader name='python/triton/compiler/errors.py'>
<_io.BufferedReader name='python/triton/compiler/make_launcher.py'>
<_io.BufferedReader name='python/triton/backends/driver.py'>
<_io.BufferedReader name='python/triton/backends/nvidia/__init__.py'>

Separate:
<_io.BufferedReader name='python/triton/compiler/code_generator.py'>
<_io.BufferedReader name='python/triton/compiler/compiler.py'>
<_io.BufferedReader name='python/triton/compiler/errors.py'>
<_io.BufferedReader name='python/triton/compiler/make_launcher.py'>
**<_io.BufferedReader name='python/triton/backends/compiler.py'>**
<_io.BufferedReader name='python/triton/backends/driver.py'>
<_io.BufferedReader name=python/triton/backends/nvidia/__init__.py'>

Switching from 
   `pkgutil.iter_modules(paths)`
to
  ```
for path in paths:
    pkgutil.iter_modules(path)
```
just worked.

Reply:
Hm, I suspect there's something we don't understand about how iter_modules works, rather than there being a bug in such a fundamental Python library.

It's probably worth engaging with upstream to understand what the expected behavior here is, because if there's something we don't understand, we may still have something wrong in the current implementation.

Reply:
There's some pretty unintuitive behavior in `iter_modules`.  It will only yield each name once (https://github.com/python/cpython/blob/3.12/Lib/pkgutil.py#L117) but we have `compiler.py` in multiple submodules.

It seems like `pkgutil` expects to be told module prefixes, which is not totally surprising since it's a fairly low level library. This is what I would do:
```
    path_prefixes = [
        (os.path.join(TRITON_PATH, "compiler"), "triton.compiler."),
        (os.path.join(TRITON_PATH, "backends"), "triton.backends."),
    ]
    for path, prefix in path_prefixes:
        for lib in pkgutil.walk_packages([path], prefix=prefix):
            with open(lib.module_finder.find_spec(lib.name).origin, "rb") as f:
                contents += [hashlib.sha256(f.read()).hexdigest()]
```

Reply:
Thanks for pointing out the prefix issue. I guess adding prefixes isn't very immune to future source changes? How does the current implementation which is based on a hand-written recursive search look to you?

Reply:
I really want there to be a clean way to do this with the standard library but I haven't been able to find one other than the prefix solution.  At least we understand why the stdlib doesn't quite do the right thing... maybe add a defensive comment to `iter_modules_recursively` and I'm good with this approach

Reply:
> I guess adding prefixes isn't very immune to future source changes?

If the fact that we have to add prefixes is the only problem with what Bert wrote above, then I think we should go with that, because it's simpler.  The fact is, this is already somewhat fragile because we have to list the directories we want to track -- this fragility is what created the need for this PR, but this PR doesn't fix the fragility (and I don't see how it could).  So adding a bit more fragility is not a big deal IMO.  OTOH Bert's implementation avoids writing a custom recursive-search function, which is nice.

In fact I don't think it should be a huge deal to hash the standard library.  You could make an argument that we really *should* hash these dependencies.  The standard library is not likely to change, but e.g. if this part of Triton were to grow a dependency on other libraries, then we really *should* hash them.

Anyway just my 2c, I will approve either approach.  :)

Reply:
Good point. Let's use Bert's implementation.